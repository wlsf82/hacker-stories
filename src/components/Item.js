import React from 'react';

export const Item = ({ item }) =>
  <div>
    <span>
      <a href={item.url}>{item.title}</a>
    </span>
    <span>{item.author}</span>
    <span>{item.num_comments}</span>
    <span>{item.points}</span>
  </div>
