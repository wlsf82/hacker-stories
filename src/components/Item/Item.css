.item {
  display: flex;
  align-items: center;
  padding-bottom: 5px;
}

.item > span {
  padding: 0 5px;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
}

.item > span > a {
  color: inherit;
}
