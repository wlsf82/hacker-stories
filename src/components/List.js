import React from 'react';

const list = [
  {
    title: 'React',
    url: 'https://reactjs.org/',
    author: 'Jordan Walke',
    num_comments: 3,
    points: 4,
    objectID: 0,
  },
  {
    title: 'Redux',
    url: 'https://redux.js.org/',
    author: 'Dan Abramov, Andrew Clark',
    num_comments: 2,
    points: 5,
    objectID: 1,
  },
];

export const List = () =>
  list.map(item => (
    <div key={item.objectID}>
      <span>
        <a href={item.url}>{item.title}</a>
      </span>
      <span>{item.author}</span>
      <span>{item.num_comments}</span>
      <span>{item.points}</span>
    </div>
  )
);
