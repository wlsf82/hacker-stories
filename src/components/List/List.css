.item-header > span {
  padding: 5px 5px;
}