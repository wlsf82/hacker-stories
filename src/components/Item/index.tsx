import React from 'react';

import { Button } from '../Button';

import { ReactComponent as Check } from '../../check.svg';

import { ItemProps } from '../../types';

import './Item.css';

export const Item = ({ item, onRemoveItem } : ItemProps) => (
  <div className="item">
    <span style={{ width: '40%' }}>
      <a
        href={item.url}
        target="_blank"
        rel="noopener noreferrer"
      >
        {item.title}
      </a>
    </span>
    <span style={{ width: '30%' }}>{item.author}</span>
    <span style={{ width: '10%' }}>{item.num_comments}</span>
    <span style={{ width: '10%' }}>{item.points}</span>
    <span style={{ width: '10%' }}>
      <Button
        onClick={() => onRemoveItem(item)}
        className="button button-small"
      >
        <Check height="18px" width="18px" />
      </Button>
    </span>
  </div>
);
