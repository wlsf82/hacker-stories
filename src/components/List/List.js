import React from 'react';

import { Item } from '../Item/Item';

export const List = ({ list, onRemoveItem }) =>
  list.map(item => (
    <Item
      key={item.objectID}
      item={item}
      onRemoveItem={onRemoveItem}
    />
  ));
