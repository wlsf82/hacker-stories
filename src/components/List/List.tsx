import React from 'react';

import { Item } from '../Item/Item';

import { ListProps } from '../../types';

export const List = ({ list, onRemoveItem }: ListProps) => (
  <>
    {list.map(item => (
      <Item
        key={item.objectID}
        item={item}
        onRemoveItem={onRemoveItem}
      />
    ))}
  </>
);
