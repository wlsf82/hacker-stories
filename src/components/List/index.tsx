import React from 'react';
import { sortBy } from 'lodash';

import Button from '../Button';
import Item from '../Item';

import { ListProps, Story, Stories } from '../../types';

import './List.css';

enum SORTS_ENUM {
  NONE = 'NONE',
  TITLE = 'TITLE',
  AUTHOR = 'AUTHOR',
  COMMENT = 'COMMENT',
  POINT = 'POINT',
}

const SORTS = {
  NONE: (list: Stories) => list,
  TITLE: (list: Stories) => sortBy(list, 'title'),
  AUTHOR: (list: Stories) => sortBy(list, 'author'),
  COMMENT: (list: Stories) => sortBy(list, 'num_comments').reverse(),
  POINT: (list: Stories) => sortBy(list, 'points').reverse(),
};

const List = ({ list, onRemoveItem }: ListProps) => {
  const [sort, setSort] = React.useState<SORTS_ENUM>(SORTS_ENUM.NONE);

  const handleSort = (sortKey: SORTS_ENUM) => {
    setSort(sortKey);
  };

  const sortFunction = SORTS[sort];
  const sortedList = sortFunction(list);

  return (
    <div>
      <div className="item-header" style={{ display: 'flex' }}>
        <span style={{ width: '40%' }}>
          <Button onClick={() => handleSort(SORTS_ENUM.TITLE)}>
            Title
          </Button>
        </span>
        <span style={{ width: '30%' }}>
          <Button onClick={() => handleSort(SORTS_ENUM.AUTHOR)}>
            Author
          </Button>
        </span>
        <span style={{ width: '10%' }}>
          <Button onClick={() => handleSort(SORTS_ENUM.COMMENT)}>
            Comments
          </Button>
        </span>
        <span style={{ width: '10%' }}>
          <Button onClick={() => handleSort(SORTS_ENUM.POINT)}>
            Points
          </Button>
        </span>
        <span style={{ width: '10%' }}>
          Action
        </span>
      </div>

      {sortedList.map((item: Story) => (
        <Item
          key={item.objectID}
          item={item}
          onRemoveItem={onRemoveItem}
        />
      ))}
    </div>
  );
};

export default List;
