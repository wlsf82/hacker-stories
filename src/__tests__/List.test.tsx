import React from 'react';
import renderer from 'react-test-renderer';

import Item from '../components/Item';
import List from '../components/List';

describe('List', () => {
  const list = [
    {
      title: 'React',
      url: 'https://reactjs.org/',
      author: 'Jordan Walke',
      num_comments: 3,
      points: 4,
      objectID: 0,
    },
    {
      title: 'Redux',
      url: 'https://redux.js.org/',
      author: 'Dan Abramov, Andrew Clark',
      num_comments: 2,
      points: 5,
      objectID: 1,
    },
  ];

  const handleRemoveItem = jest.fn();

  let component: any;

  beforeEach(() => {
    component = renderer.create(<List list={list} onRemoveItem={handleRemoveItem} />);
  });

  it('renders two items', () => {
    expect(component.root.findAllByType(Item).length).toEqual(2);
  });

  it('renders snapshot', () => {
    let tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });
});
