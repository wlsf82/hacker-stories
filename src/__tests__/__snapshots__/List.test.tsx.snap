// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List renders snapshot 1`] = `
<div>
  <div
    className="item-header"
    style={
      Object {
        "display": "flex",
      }
    }
  >
    <span
      style={
        Object {
          "width": "40%",
        }
      }
    >
      <button
        className=""
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Title
      </button>
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
      <button
        className=""
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Author
      </button>
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className=""
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Comments
      </button>
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className=""
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Points
      </button>
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      Action
    </span>
  </div>
  <div
    className="item"
    style={
      Object {
        "display": "flex",
      }
    }
  >
    <span
      style={
        Object {
          "width": "40%",
        }
      }
    >
      <a
        href="https://reactjs.org/"
        rel="noopener noreferrer"
        target="_blank"
      >
        React
      </a>
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
      Jordan Walke
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      3
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      4
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className="button button-small"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        <svg
          height="18px"
          width="18px"
        >
          check.svg
        </svg>
      </button>
    </span>
  </div>
  <div
    className="item"
    style={
      Object {
        "display": "flex",
      }
    }
  >
    <span
      style={
        Object {
          "width": "40%",
        }
      }
    >
      <a
        href="https://redux.js.org/"
        rel="noopener noreferrer"
        target="_blank"
      >
        Redux
      </a>
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
      Dan Abramov, Andrew Clark
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      2
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      5
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className="button button-small"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        <svg
          height="18px"
          width="18px"
        >
          check.svg
        </svg>
      </button>
    </span>
  </div>
</div>
`;
